backends/p4tools/p4tools.def

```diff
@@ -117,10 +126,83 @@ class StateVariable : Expression {
         return 0;
     }
 
+    uint64_t hash() const {
+        return hash(0, ref);
+    }
+
+    uint64_t hash(uint64_t seed, const IR::Expression *expression) const {
+        // expression is a Member.
+        if (const auto *member = expression->to<IR::Member>()) {
+            return hash(seed, member);
+        }
+        // expression is a PathExpression.
+        if (const auto *pathExpression = expression->to<IR::PathExpression>()) {
+            return hash(seed, pathExpression);
+        }
+        // expression is a ArrayIndex.
+        if (const auto *arrayIndex = expression->to<IR::ArrayIndex>()) {
+            return hash(seed, arrayIndex);
+        }
+        BUG("Either %1% of type %2% is not a valid StateVariable", expression,
+            expression->node_type_name());
+    }
+
+    uint64_t hash(uint64_t seed, const IR::Member *member) const {
+        return hash(Util::hash_combine(seed, std::hash<cstring>()(member->member)), member->expr);
+    }
+
+    uint64_t hash(uint64_t seed, const IR::PathExpression *pathExpression) const {
+        return Util::hash_combine(seed, std::hash<cstring>()(pathExpression->path->name));
+    }
+
+    uint64_t hash(uint64_t seed, const IR::ArrayIndex *arrayIndex) const {
+        const auto *arrayIndexVal = arrayIndex->right->to<IR::Constant>();
+        BUG_CHECK(
+            arrayIndexVal != nullptr,
+            "Value %1% is not a constant. Only constants are supported as part of a state variable.",
+            arrayIndex->right);
+        return hash(Util::hash_combine(seed, std::hash<big_int>()(arrayIndexVal->value)),
+                                 arrayIndex->left);
+    }
+
```

Comment:
I wonder if it would be better to implement the hash outside the IR classes. Having the multiple public overloads there on the public interface seems messy to me.

Reply:
My problem with that approach is that it can be tedious to find the implementation. But in this case it makes sense, I just need to implement a new class.

Reply:
> But in this case it makes sense, I just need to implement a new class.

Any updates on this?

Reply:
This PR is still blocked until #4768 is resolved.